File: mycelium/languages/c_cpp.py
"""C and C++ language analysers."""

from __future__ import annotations

import tree_sitter
import tree_sitter_c as ts_c
import tree_sitter_cpp as ts_cpp

from mycelium.config import ImportStatement, RawCall, Symbol, SymbolType, Visibility


class _CBaseMixin:
    """Shared functionality for C and C++ analysers."""

    _PREPROC_CONTAINERS = {"preproc_ifdef", "preproc_ifndef", "preproc_if", "preproc_else", "preproc_elif"}

    def _extract_symbols_from_node(self, node, source, file_path, symbols, parent_id=None):
        for child in node.children:
            if child.type == "function_definition":
                name = self._get_func_name(child)
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.FUNCTION,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
            elif child.type == "struct_specifier":
                name = self._get_type_name(child)
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.STRUCT,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
            elif child.type == "enum_specifier":
                name = self._get_type_name(child)
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.ENUM,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
            elif child.type == "type_definition":
                name = self._get_typedef_name(child)
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.TYPEDEF,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
            elif child.type == "declaration":
                # Forward declarations of functions
                name = self._get_func_name(child)
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.FUNCTION,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
            elif child.type in self._PREPROC_CONTAINERS:
                self._extract_symbols_from_node(child, source, file_path, symbols, parent_id)

    def _get_func_name(self, node) -> str | None:
        for child in node.children:
            if child.type == "function_declarator":
                for c in child.children:
                    if c.type == "identifier":
                        return c.text.decode("utf-8")
            if child.type == "pointer_declarator":
                result = self._get_func_name(child)
                if result:
                    return result
            if child.type == "identifier":
                return child.text.decode("utf-8")
        return None

    def _get_type_name(self, node) -> str | None:
        for child in node.children:
            if child.type == "type_identifier":
                return child.text.decode("utf-8")
        return None

    def _get_typedef_name(self, node) -> str | None:
        for child in node.children:
            if child.type == "type_identifier":
                return child.text.decode("utf-8")
        return None

    def _extract_includes(self, tree, file_path):
        imports = []
        for child in tree.root_node.children:
            if child.type == "preproc_include":
                path = None
                for c in child.children:
                    if c.type == "string_literal":
                        for sc in c.children:
                            if sc.type == "string_content":
                                path = sc.text.decode("utf-8")
                    elif c.type == "system_lib_string":
                        path = c.text.decode("utf-8").strip("<>")
                if path:
                    imports.append(ImportStatement(
                        file=file_path,
                        statement=child.text.decode("utf-8").strip(),
                        target_name=path,
                        line=child.start_point[0] + 1,
                    ))
        return imports

    def _find_call_expressions(self, node, file_path, calls, exclusions):
        if node.type == "call_expression":
            callee_name, qualifier = self._extract_callee(node)
            if callee_name and callee_name not in exclusions:
                qualified = f"{qualifier}.{callee_name}" if qualifier else callee_name
                if qualified not in exclusions:
                    caller = self._find_enclosing_func(node)
                    calls.append(RawCall(
                        caller_file=file_path,
                        caller_name=caller or "<module>",
                        callee_name=callee_name,
                        line=node.start_point[0] + 1,
                        qualifier=qualifier,
                    ))
        for child in node.children:
            self._find_call_expressions(child, file_path, calls, exclusions)

    def _extract_callee(self, node):
        first = node.children[0] if node.children else None
        if first is None:
            return None, None
        if first.type == "identifier":
            return first.text.decode("utf-8"), None
        if first.type == "field_expression":
            parts = []
            for c in first.children:
                if c.type in ("identifier", "field_identifier"):
                    parts.append(c.text.decode("utf-8"))
            if len(parts) >= 2:
                return parts[-1], parts[-2]
            elif parts:
                return parts[0], None
        if first.type == "qualified_identifier":
            text = first.text.decode("utf-8")
            parts = text.rsplit("::", 1)
            if len(parts) == 2:
                return parts[1], parts[0]
            return text, None
        return None, None

    def _find_enclosing_func(self, node):
        current = node.parent
        while current:
            if current.type == "function_definition":
                return self._get_func_name(current)
            current = current.parent
        return None


class CAnalyser(_CBaseMixin):
    extensions = [".c", ".h"]
    language_name = "c"

    def get_language(self) -> tree_sitter.Language:
        return tree_sitter.Language(ts_c.language())

    def extract_symbols(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[Symbol]:
        symbols: list[Symbol] = []
        self._extract_symbols_from_node(tree.root_node, source, file_path, symbols)
        return symbols

    def extract_imports(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[ImportStatement]:
        return self._extract_includes(tree, file_path)

    def extract_calls(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[RawCall]:
        calls: list[RawCall] = []
        self._find_call_expressions(tree.root_node, file_path, calls, self.builtin_exclusions())
        return calls

    def builtin_exclusions(self) -> set[str]:
        return {
            "printf", "fprintf", "sprintf", "snprintf", "scanf", "sscanf",
            "malloc", "calloc", "realloc", "free",
            "memcpy", "memset", "memmove", "memcmp",
            "strlen", "strcmp", "strncmp", "strcpy", "strncpy", "strcat",
            "sizeof", "assert", "exit", "abort",
            "fopen", "fclose", "fread", "fwrite", "fgets", "fputs",
        }


class CppAnalyser(_CBaseMixin):
    extensions = [".cpp", ".cc", ".cxx", ".hpp", ".hxx", ".hh"]
    language_name = "cpp"

    def get_language(self) -> tree_sitter.Language:
        return tree_sitter.Language(ts_cpp.language())

    def extract_symbols(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[Symbol]:
        symbols: list[Symbol] = []
        self._extract_cpp_symbols(tree.root_node, source, file_path, symbols, parent_id=None)
        return symbols

    def _extract_cpp_symbols(self, node, source, file_path, symbols, parent_id):
        self._extract_symbols_from_node(node, source, file_path, symbols, parent_id)

        for child in node.children:
            if child.type == "class_specifier":
                name = self._get_type_name(child)
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.CLASS,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
            elif child.type == "namespace_definition":
                name = None
                for c in child.children:
                    if c.type == "namespace_identifier":
                        name = c.text.decode("utf-8")
                        break
                if name:
                    symbols.append(Symbol(
                        id=f"_pending_{len(symbols)}",
                        name=name,
                        type=SymbolType.NAMESPACE,
                        file=file_path,
                        line=child.start_point[0] + 1,
                        visibility=Visibility.PUBLIC,
                        exported=True,
                        parent=parent_id,
                    ))
                    # Recurse into namespace
                    for c in child.children:
                        if c.type == "declaration_list":
                            self._extract_cpp_symbols(c, source, file_path, symbols, parent_id=name)

    def extract_imports(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[ImportStatement]:
        return self._extract_includes(tree, file_path)

    def extract_calls(
        self, tree: tree_sitter.Tree, source: bytes, file_path: str
    ) -> list[RawCall]:
        calls: list[RawCall] = []
        self._find_call_expressions(tree.root_node, file_path, calls, self.builtin_exclusions())
        return calls

    def builtin_exclusions(self) -> set[str]:
        return {
            "printf", "malloc", "free", "memcpy", "memset", "strlen", "strcmp",
            "sizeof", "assert", "exit", "abort",
            "std::cout", "std::cerr", "std::endl",
            "std::move", "std::forward",
            "std::make_shared", "std::make_unique", "std::make_pair",
            "std::sort", "std::find", "std::begin", "std::end",
            "std::string", "std::to_string",
        }

File: crates/mycelium-core/src/languages/c_cpp.rs
//! C and C++ language analysers (shared implementation).

use std::collections::HashSet;
use std::sync::LazyLock;

use tree_sitter::{Language, Node, Tree};

use super::LanguageAnalyser;
use crate::config::{ImportStatement, RawCall, Symbol, SymbolType, Visibility};

static C_BUILTIN_EXCLUSIONS: LazyLock<HashSet<String>> = LazyLock::new(|| {
    [
        "printf", "fprintf", "sprintf", "snprintf", "scanf", "fscanf", "sscanf", "malloc",
        "calloc", "realloc", "free", "memcpy", "memmove", "memset", "memcmp", "strlen", "strcpy",
        "strncpy", "strcat", "strncat", "strcmp", "strncmp", "fopen", "fclose", "fread", "fwrite",
        "fgets", "fputs", "exit", "abort", "atexit", "assert", "sizeof", "offsetof",
    ]
    .into_iter()
    .map(String::from)
    .collect()
});

static CPP_BUILTIN_EXCLUSIONS: LazyLock<HashSet<String>> = LazyLock::new(|| {
    let mut set = C_BUILTIN_EXCLUSIONS.clone();
    for name in [
        "std::cout",
        "std::cerr",
        "std::endl",
        "std::make_shared",
        "std::make_unique",
        "std::make_pair",
        "std::move",
        "std::forward",
        "std::swap",
        "std::sort",
        "std::find",
        "std::transform",
        "std::begin",
        "std::end",
        "std::string",
        "std::to_string",
        "std::stoi",
        "std::stof",
        "std::vector",
        "std::map",
        "std::set",
        "std::unordered_map",
        "static_cast",
        "dynamic_cast",
        "reinterpret_cast",
        "const_cast",
        "new",
        "delete",
    ] {
        set.insert(name.to_string());
    }
    set
});

const PREPROC_CONTAINERS: &[&str] = &[
    "preproc_ifdef",
    "preproc_ifndef",
    "preproc_if",
    "preproc_else",
    "preproc_elif",
];

fn is_preproc_container(kind: &str) -> bool {
    PREPROC_CONTAINERS.contains(&kind)
}

// ---- Shared C/C++ helpers ----

fn get_func_name(node: &Node, source: &[u8]) -> Option<String> {
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if child.kind() == "function_declarator" {
                for j in 0..child.child_count() {
                    if let Some(c) = child.child(j) {
                        if c.kind() == "identifier" {
                            return c.utf8_text(source).ok().map(|s| s.to_string());
                        }
                    }
                }
            }
            if child.kind() == "pointer_declarator" {
                let result = get_func_name(&child, source);
                if result.is_some() {
                    return result;
                }
            }
            if child.kind() == "identifier" {
                return child.utf8_text(source).ok().map(|s| s.to_string());
            }
        }
    }
    None
}

fn get_type_name(node: &Node, source: &[u8]) -> Option<String> {
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if child.kind() == "type_identifier" {
                return child.utf8_text(source).ok().map(|s| s.to_string());
            }
        }
    }
    None
}

fn extract_c_symbols(
    node: &Node,
    source: &[u8],
    file_path: &str,
    symbols: &mut Vec<Symbol>,
    parent_id: Option<&str>,
    lang: &str,
) {
    for i in 0..node.child_count() {
        let child = match node.child(i) {
            Some(c) => c,
            None => continue,
        };

        if child.kind() == "function_definition" {
            if let Some(name) = get_func_name(&child, source) {
                symbols.push(Symbol {
                    id: format!("_pending_{}", symbols.len()),
                    name,
                    symbol_type: SymbolType::Function,
                    file: file_path.to_string(),
                    line: child.start_position().row + 1,
                    visibility: Visibility::Public,
                    exported: true,
                    parent: parent_id.map(|s| s.to_string()),
                    language: Some(lang.to_string()),
                    byte_range: Some((child.byte_range().start, child.byte_range().end)),
                    parameter_types: None,
                });
            }
        } else if child.kind() == "struct_specifier" {
            if let Some(name) = get_type_name(&child, source) {
                symbols.push(Symbol {
                    id: format!("_pending_{}", symbols.len()),
                    name,
                    symbol_type: SymbolType::Struct,
                    file: file_path.to_string(),
                    line: child.start_position().row + 1,
                    visibility: Visibility::Public,
                    exported: true,
                    parent: parent_id.map(|s| s.to_string()),
                    language: Some(lang.to_string()),
                    byte_range: Some((child.byte_range().start, child.byte_range().end)),
                    parameter_types: None,
                });
            }
        } else if child.kind() == "enum_specifier" {
            if let Some(name) = get_type_name(&child, source) {
                symbols.push(Symbol {
                    id: format!("_pending_{}", symbols.len()),
                    name,
                    symbol_type: SymbolType::Enum,
                    file: file_path.to_string(),
                    line: child.start_position().row + 1,
                    visibility: Visibility::Public,
                    exported: true,
                    parent: parent_id.map(|s| s.to_string()),
                    language: Some(lang.to_string()),
                    byte_range: Some((child.byte_range().start, child.byte_range().end)),
                    parameter_types: None,
                });
            }
        } else if child.kind() == "type_definition" {
            // typedef
            if let Some(name) = get_type_name(&child, source) {
                symbols.push(Symbol {
                    id: format!("_pending_{}", symbols.len()),
                    name,
                    symbol_type: SymbolType::Typedef,
                    file: file_path.to_string(),
                    line: child.start_position().row + 1,
                    visibility: Visibility::Public,
                    exported: true,
                    parent: parent_id.map(|s| s.to_string()),
                    language: Some(lang.to_string()),
                    byte_range: Some((child.byte_range().start, child.byte_range().end)),
                    parameter_types: None,
                });
            }
        } else if child.kind() == "declaration" {
            // Forward declarations of functions
            if let Some(name) = get_func_name(&child, source) {
                symbols.push(Symbol {
                    id: format!("_pending_{}", symbols.len()),
                    name,
                    symbol_type: SymbolType::Function,
                    file: file_path.to_string(),
                    line: child.start_position().row + 1,
                    visibility: Visibility::Public,
                    exported: true,
                    parent: parent_id.map(|s| s.to_string()),
                    language: Some(lang.to_string()),
                    byte_range: Some((child.byte_range().start, child.byte_range().end)),
                    parameter_types: None,
                });
            }
        } else if is_preproc_container(child.kind()) {
            extract_c_symbols(&child, source, file_path, symbols, parent_id, lang);
        }
    }
}

fn extract_includes(tree: &Tree, source: &[u8], file_path: &str) -> Vec<ImportStatement> {
    let mut imports = Vec::new();
    let root = tree.root_node();
    for i in 0..root.child_count() {
        if let Some(child) = root.child(i) {
            if child.kind() == "preproc_include" {
                let mut path = None;
                for j in 0..child.child_count() {
                    if let Some(c) = child.child(j) {
                        if c.kind() == "string_literal" {
                            for k in 0..c.child_count() {
                                if let Some(sc) = c.child(k) {
                                    if sc.kind() == "string_content" {
                                        path = sc.utf8_text(source).ok().map(|s| s.to_string());
                                    }
                                }
                            }
                        } else if c.kind() == "system_lib_string" {
                            if let Ok(text) = c.utf8_text(source) {
                                path =
                                    Some(text.trim_matches(|c| c == '<' || c == '>').to_string());
                            }
                        }
                    }
                }
                if let Some(path) = path {
                    imports.push(ImportStatement {
                        file: file_path.to_string(),
                        statement: child.utf8_text(source).unwrap_or("").trim().to_string(),
                        target_name: path,
                        line: child.start_position().row + 1,
                    });
                }
            }
        }
    }
    imports
}

fn find_c_calls(
    node: &Node,
    source: &[u8],
    file_path: &str,
    calls: &mut Vec<RawCall>,
    exclusions: &HashSet<String>,
) {
    if node.kind() == "call_expression" {
        let (callee_name, qualifier) = extract_c_callee(node, source);
        if let Some(ref name) = callee_name {
            if !exclusions.contains(name) {
                let qualified = if let Some(ref q) = qualifier {
                    format!("{}.{}", q, name)
                } else {
                    name.clone()
                };
                if !exclusions.contains(&qualified) {
                    let caller = find_enclosing_func(node, source);
                    calls.push(RawCall {
                        caller_file: file_path.to_string(),
                        caller_name: caller.unwrap_or_else(|| "<module>".to_string()),
                        callee_name: name.clone(),
                        line: node.start_position().row + 1,
                        qualifier,
                    });
                }
            }
        }
    }
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            find_c_calls(&child, source, file_path, calls, exclusions);
        }
    }
}

fn extract_c_callee(node: &Node, source: &[u8]) -> (Option<String>, Option<String>) {
    let first = match node.child(0) {
        Some(c) => c,
        None => return (None, None),
    };

    if first.kind() == "identifier" {
        return (first.utf8_text(source).ok().map(|s| s.to_string()), None);
    }

    if first.kind() == "field_expression" {
        let mut parts = Vec::new();
        for i in 0..first.child_count() {
            if let Some(c) = first.child(i) {
                if c.kind() == "identifier" || c.kind() == "field_identifier" {
                    if let Ok(text) = c.utf8_text(source) {
                        parts.push(text.to_string());
                    }
                }
            }
        }
        if parts.len() >= 2 {
            let name = parts.pop();
            let qualifier = parts.pop();
            return (name, qualifier);
        } else if !parts.is_empty() {
            return (Some(parts.remove(0)), None);
        }
    }

    if first.kind() == "qualified_identifier" {
        if let Ok(text) = first.utf8_text(source) {
            let text = text.to_string();
            if let Some(pos) = text.rfind("::") {
                return (
                    Some(text[pos + 2..].to_string()),
                    Some(text[..pos].to_string()),
                );
            }
            return (Some(text), None);
        }
    }

    (None, None)
}

fn find_enclosing_func(node: &Node, source: &[u8]) -> Option<String> {
    let mut current = node.parent();
    while let Some(n) = current {
        if n.kind() == "function_definition" {
            return get_func_name(&n, source);
        }
        current = n.parent();
    }
    None
}

// ---- C Analyser ----

/// C language analyser.
pub struct CAnalyser;

impl Default for CAnalyser {
    fn default() -> Self {
        Self
    }
}

impl CAnalyser {
    pub fn new() -> Self {
        Self
    }
}

impl LanguageAnalyser for CAnalyser {
    fn extensions(&self) -> &[&str] {
        &["c", "h"]
    }

    fn language_name(&self) -> &str {
        "C"
    }

    fn get_language(&self) -> Language {
        tree_sitter_c::LANGUAGE.into()
    }

    fn extract_symbols(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<Symbol> {
        let mut symbols = Vec::new();
        extract_c_symbols(
            &tree.root_node(),
            source,
            file_path,
            &mut symbols,
            None,
            "C",
        );
        symbols
    }

    fn extract_imports(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<ImportStatement> {
        extract_includes(tree, source, file_path)
    }

    fn extract_calls(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<RawCall> {
        let mut calls = Vec::new();
        let exclusions = self.builtin_exclusions();
        find_c_calls(&tree.root_node(), source, file_path, &mut calls, exclusions);
        calls
    }

    fn builtin_exclusions(&self) -> &HashSet<String> {
        &C_BUILTIN_EXCLUSIONS
    }
}

// ---- C++ Analyser ----

/// C++ language analyser.
pub struct CppAnalyser;

impl Default for CppAnalyser {
    fn default() -> Self {
        Self
    }
}

impl CppAnalyser {
    pub fn new() -> Self {
        Self
    }

    fn extract_cpp_symbols(
        &self,
        node: &Node,
        source: &[u8],
        file_path: &str,
        symbols: &mut Vec<Symbol>,
        parent_id: Option<&str>,
    ) {
        // Extract base C symbols (functions, structs, enums, typedefs)
        extract_c_symbols(node, source, file_path, symbols, parent_id, "C++");

        // C++-specific: classes and namespaces
        for i in 0..node.child_count() {
            let child = match node.child(i) {
                Some(c) => c,
                None => continue,
            };

            if child.kind() == "class_specifier" {
                if let Some(name) = get_type_name(&child, source) {
                    symbols.push(Symbol {
                        id: format!("_pending_{}", symbols.len()),
                        name,
                        symbol_type: SymbolType::Class,
                        file: file_path.to_string(),
                        line: child.start_position().row + 1,
                        visibility: Visibility::Public,
                        exported: true,
                        parent: parent_id.map(|s| s.to_string()),
                        language: Some("C++".to_string()),
                        byte_range: Some((child.byte_range().start, child.byte_range().end)),
                        parameter_types: None,
                    });
                }
            } else if child.kind() == "namespace_definition" {
                let mut name = None;
                for j in 0..child.child_count() {
                    if let Some(c) = child.child(j) {
                        if c.kind() == "namespace_identifier" {
                            name = c.utf8_text(source).ok().map(|s| s.to_string());
                            break;
                        }
                    }
                }
                if let Some(ref ns_name) = name {
                    symbols.push(Symbol {
                        id: format!("_pending_{}", symbols.len()),
                        name: ns_name.clone(),
                        symbol_type: SymbolType::Namespace,
                        file: file_path.to_string(),
                        line: child.start_position().row + 1,
                        visibility: Visibility::Public,
                        exported: true,
                        parent: parent_id.map(|s| s.to_string()),
                        language: Some("C++".to_string()),
                        byte_range: Some((child.byte_range().start, child.byte_range().end)),
                        parameter_types: None,
                    });
                    // Recurse into namespace body
                    for j in 0..child.child_count() {
                        if let Some(c) = child.child(j) {
                            if c.kind() == "declaration_list" {
                                self.extract_cpp_symbols(
                                    &c,
                                    source,
                                    file_path,
                                    symbols,
                                    Some(ns_name),
                                );
                            }
                        }
                    }
                }
            }
        }
    }
}

impl LanguageAnalyser for CppAnalyser {
    fn extensions(&self) -> &[&str] {
        &["cpp", "cxx", "cc", "hpp", "hxx", "hh"]
    }

    fn language_name(&self) -> &str {
        "C++"
    }

    fn get_language(&self) -> Language {
        tree_sitter_cpp::LANGUAGE.into()
    }

    fn extract_symbols(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<Symbol> {
        let mut symbols = Vec::new();
        self.extract_cpp_symbols(&tree.root_node(), source, file_path, &mut symbols, None);
        symbols
    }

    fn extract_imports(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<ImportStatement> {
        extract_includes(tree, source, file_path)
    }

    fn extract_calls(&self, tree: &Tree, source: &[u8], file_path: &str) -> Vec<RawCall> {
        let mut calls = Vec::new();
        let exclusions = self.builtin_exclusions();
        find_c_calls(&tree.root_node(), source, file_path, &mut calls, exclusions);
        calls
    }

    fn builtin_exclusions(&self) -> &HashSet<String> {
        &CPP_BUILTIN_EXCLUSIONS
    }
}
